internal/proxy/gsettings.go

```diff
@@ -0,0 +1,176 @@
+			}
+		}
+	case protocolNo:
+		// Ignored hosts are configured at the root level
+		section = fmt.Sprintf("[%s]", systemProxySchemaID)
+
+		hosts := strings.Split(p.escapedURL, ",")
+		for i, host := range hosts {
+			hosts[i] = wrapHostIfNeeded(host)
+		}
+		settings = fmt.Sprintf("ignore-hosts=[%s]\n", strings.Join(hosts, ","))
+	case protocolAuto:
+		// Autoconfig URL is configured at the root level
+		section = fmt.Sprintf("[%s]", systemProxySchemaID)
+		settings = fmt.Sprintf("autoconfig-url='%s'\n", p.escapedURL)
+	}
+
+	return fmt.Sprintf("%s\n%s\n", section, settings)
+}
+
+// applyToGSettings applies the proxy configuration in the form of a GSchema override file,
+// then runs glib-compile-schemas to make the changes visible to GSettings.
+// If there are no proxy settings to apply, the GSchema override file is removed.
+func (p Proxy) applyToGSettings() (err error) {
+	defer decorate.OnError(&err, "couldn't apply GSettings proxy configuration")
+
+	// On the off chance that the user is not running GNOME, we want to print a warning and quietly return.
+	if _, err := exec.LookPath(p.glibCompileSchemasCmd[0]); err != nil {
+		log.Warningf("Couldn't find an executable for %q, not applying GSettings proxy configuration", p.glibCompileSchemasCmd[0])
+		return nil
+	}
+
+	// Check if the parent directory exists - fail if it doesn't, as it means we
+	// don't have any defined proxy XML schema to override.
+	if stat, err := os.Stat(filepath.Dir(p.gsettingsConfigPath)); err != nil {
+		return fmt.Errorf("couldn't find GLib schema directory: %w", err)
+	} else if !stat.IsDir() {
+		return fmt.Errorf("GLib schema path %q is not a directory", filepath.Dir(p.gsettingsConfigPath))
+	}
+
+	if len(p.settings) == 0 {
+		log.Debug("No proxy settings to apply, removing GSchema override file if it exists")
+
+		// If we managed to remove something, we need to recompile the schemas
+		// to propagate the change to GSettings.
+		if err := os.Remove(p.gsettingsConfigPath); err == nil {
+			log.Debugf("Removed GSettings override file at %q", p.gsettingsConfigPath)
+			return p.runGlibCompileSchemas()
+		} else if !errors.Is(err, fs.ErrNotExist) {
+			return err
+		}
+		return nil
+	}
+	log.Debugf("Applying GSettings proxy configuration to %q", p.gsettingsConfigPath)
+
+	content := p.gsettingsConfig()
+	prevContent, err := previousConfig(p.gsettingsConfigPath)
+	if err == nil && prevContent == content {
+		log.Debugf("GSettings proxy configuration at %q is already up to date", p.gsettingsConfigPath)
+		return nil
+	} else if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return err
+	}
+
+	if err := safeWriteFile(p.gsettingsConfigPath, content); err != nil {
+		return err
+	}
+
+	if err := p.runGlibCompileSchemas(); err != nil {
+		// If we failed to recompile the schemas (due to our fault or not),
+		// revert to the previous version of the configuration file.
+		return errors.Join(err, safeWriteFile(p.gsettingsConfigPath, prevContent))
+	}
```

Comment:
If we need a restore point in case we need to restore the schemas, maybe it's better to, instead of overwriting the file in the first place, rename the previous one to `config.old` or `config.prev` and then restore it to `config` if we fail to apply the new one, wdyt? It will take some refactoring, but it will take less writing operations in the end.

Ps. Maybe we can go for a bool argument in `safeWriteFile` to determine if we should backup the old file or not, wdyt?